### PR TITLE
Epic 4: Audit process and implementation for stock keepers 

### PR DESCRIPTION
**Pull request by makchamp**
This pull request touches on various aspects of user stories 4.2(#57), 4.5(#60), 4.6(#61), 4.7(#70) and 4.8(#131).

#### Audit Model
- The Audit Model has been updated to include  two sub-models (BinToSK and Record).
- The previous ItemToSK  model has been changed to BinToSK. This model will still reference its containing audit as well as the assigned stock keeper, however bins are no longer stored as a list of ids but rather represent their own instances of the BinToSK model. This new representation increases flexibility and allows us to take full advantage of the framework's builtin functionality for model view sets. 
- The Audit Model has been extended to include DateTime fields for initiated_on and last_modified_on as well as a new user reference initiated_by.
- The Record model is the implementation of a validated item. A record is made every time an item is scanned or flagged by a stock keeper.  A item can be marked, MISSING, PROVIDED or NEW. A record can be flagged as well (#60) .
- Audit and Bin models include accuracy fields. The accuracy is updated any time a record is modifed.
- Bin model now includes a status field which can be set to 'Pending' or 'Complete '
#### Action endpoints
- Added API endpoints for audits:
-  /check_item, will check if a scanned/inputted item exists in the relevant inventory
- /completed_items, can retrieve the completed items(audited) of an entire audit or particular bin
- /items will retrieve the items required to audit for a particular bin
- /progression_metrics, will return a progression report for the audit or particular bin
#### Permissions
- Reworked Stock keeper permissions to verify that a stock keeper can only retrieve information related to their own audits. A stock keeper should only be able to modify records pertaining to records or bins, not audit themselves.
#### Front-end
- Added caching service for easy access of logged in user
- Added error-handler for http requests
- Added token interceptor to allow authenticated requests
- Added loading controllers between requests for better user experience
- Removed the dashboard component (audit page is now the landing page following login)
- Removed the side navigation menu
- Reworked the toolbar to include profile and notification sections
- Added ALTA logo to the toolbar which if clicked reroutes to the Audits page
- Added Audit, Bin, Items and Record pages.
- A stock keeper can easily view all the required audits/bins/items assigned to them on their respective pages(#57 and #70). 
- On the items page the user can switch between items left and completed items (through the top segment/tabs). 
- A stock keper can delete or modify a record of an ongoing audit by clicking on any completed item.(#131)
- To audit an item, a stock keeper has 4 options. They can either use an external bluetooth scanner to scan the barcode(#205), input the barcode manually, scan it with their phone camera or simply click on the item itself. On a successful match, a modal page will be displayed presenting all the information related to the item along with options to comment (#61) on and/or flag (#60) the item. The flagging is done implicitly if the item is marked 'missing' or 'new'.  The original item is not changed, only the Audit record of the item will be different. Submitting the item will finally create a record. Items checks are being made to ensure an item is valid and the user is warned if a matched item belongs to the current audit but is from another bin.
- A stock keeper can complete a bin from the completed items page. This  sets the bins status to 'Complete'. All completed bins are listed in the Completed Bins segment on the Bins page. A bin cannot be competed if there a still items not yet completed.
- Progression metrics popover will be displayed for each Audit/Bin by clicking the 'i' info button. 
- Added coloured ellipses to audits, bins and record to visualize their current status.
#### Misc
- Configured logging for django server
Basic logging functionality has been added to the django server with old print statements converted to DEBUG and ERROR level logs
Default log level is set to WARNING however it can be changed from the env file

### CODE REVIEW

File: server/audit/views.py
Comment (NicholasNagy) [inline server/audit/views.py:125]: Why did you make this change? CheckAuditOrganizationById doesn't even check the right table in the database and now we don't check to make sure the init audit field is part of an audit that is associated with the user's ID.

With these additions, I feel like this class is vulnerable.

File: server/audit/views.py
Comment (NicholasNagy) [inline server/audit/views.py:191]: why don't you call the super class' implementation instead of rewriting the code here?

File: server/audit/views.py
Comment (NicholasNagy) [inline server/audit/views.py:163]: I don't see how these permission prevent people from creating, getting, or updating records across organizations.

File: server/audit/permissions.py
Comment (NicholasNagy) [inline server/audit/permissions.py:71]: I see the same check for whether the role is an SK. Can't we just put that in a different permission?
The goal is to make each permission as easy to understand as possible by making them small and digestible, so I think this permission alone could probably be broken up into 4, but I will let that decision go to you, but let's at least extract the check for whether the role is an SK.

**Review by NicholasNagy — CHANGES_REQUESTED** [review]
I looked over the backend, and I have a few comments related to the permissions stuff. It doesn't really follow exactly how we had planned, so please look it over when you have a chance.

I will also be submitting my review for the front end tomorrow.

File: server/audit/views.py
Comment (NicholasNagy) [reply server/audit/views.py:191]: Also, I just looked into this, you could actually just override the dispatch method instead. Would be much cleaner and you could delete all these methods.

[See this SO post for details](https://stackoverflow.com/questions/59331300/invoke-function-after-all-methods-django-rest-framework-modelviewset )

File: server/user_account/permissions.py
Comment (makchamp) [reply server/user_account/permissions.py:27]: yes maybe the factory can/should accommodate SK users

File: server/audit/views.py
Comment (makchamp) [reply server/audit/views.py:47]: its the same as the default list implementation

File: server/audit/views.py
Comment (makchamp) [reply server/audit/views.py:85]: we should add a default response just to be safe, this was done to avoid multiple return statements

File: server/audit/views.py
Comment (makchamp) [reply server/audit/views.py:99]: for sks its done with the isAssignedSK permission

File: server/audit/views.py
Comment (makchamp) [reply server/audit/views.py:108]: I want to use it across the models

File: server/audit/views.py
Comment (makchamp) [reply server/audit/views.py:152]: same as default implementation

File: server/audit/views.py
Comment (makchamp) [reply server/audit/views.py:125]: ill check out what you mean here and get back to you

File: server/audit/views.py
Comment (makchamp) [reply server/audit/views.py:191]: i want to be able to set the accuracy and maybe do other stuff before returning the response

File: server/audit/permissions.py
Comment (makchamp) [reply server/audit/permissions.py:71]: yes i agree the permissions need refactoring

File: server/audit/views.py
Comment (makchamp) [reply server/audit/views.py:163]: thats possible, i didnt change the implementation for these

File: server/user_account/permissions.py
Comment (Averynder) [reply server/user_account/permissions.py:27]: This was talked about over call, the way the factory is currently setup is meant only for SA and IMs. If We add the SKs into the factory as we did, yes we can accommodate general cases for the SK but the problem is now we have allowed for SKs to potentially gain access to other parts of our app which they wouldn't have otherwise. This isn't picked up since none of the tests test what we shouldn't have which is SK possibilities. We'll have to work on them as separate permissions unfortunatly

File: server/audit/views.py
Comment (Averynder) [reply server/audit/views.py:47]: We can take care of this through the permissions so it doesn't become a security concern and we can still use the default list

File: server/audit/views.py
Comment (Averynder) [reply server/audit/views.py:85]: We can fix this with a default case

File: server/audit/views.py
Comment (Averynder) [reply server/audit/views.py:99]: We can split this permission to make it easier to comprehend but it is handled already

File: server/audit/views.py
Comment (Averynder) [reply server/audit/views.py:152]: similar to the comment above, we can handle both with the appropriate permissions

File: server/audit/views.py
Comment (Averynder) [reply server/audit/views.py:191]: > 
> 
> i want to be able to set the accuracy and maybe do other stuff before returning the response

For anything before we'll have to have the methods defined, Nick's suggestion is simply a method that runs after all methods in the viewset, we'll have to get on call to figure out if this is what we can use optimally

File: server/audit/views.py
Comment (Averynder) [reply server/audit/views.py:163]: we'll have to work on this, the records are the latest addition to the PR, along with the rest of the comments the permissions seem to be the main thing to work on

File: server/audit/permissions.py
Comment (Averynder) [reply server/audit/permissions.py:71]: It seems this is the majority of this PR is the permissions, this will be taken care of

File: server/audit/views.py
Comment (NicholasNagy) [reply server/audit/views.py:191]: yeah, with the dispatch, it should help you do that without having to override every method. It's like a hook.

File: server/user_account/permissions.py
Comment (NicholasNagy) [reply server/user_account/permissions.py:27]: yeah exactly, these changes can be made through a new subclass that can inherit some of the functionality but provide its own custom functionality.

File: server/audit/views.py
Comment (NicholasNagy) [reply server/audit/views.py:47]: yeah exactly, we should override it to make sure we aren't using the default.

File: server/audit/views.py
Comment (NicholasNagy) [reply server/audit/views.py:85]: yeah exactly, I just found a bug where if you have 2 audits that contain the same item and a stock keeper tries to complete the item, it will raise a server error as a response. We should be able to handle that case.

File: server/audit/views.py
Comment (NicholasNagy) [reply server/audit/views.py:99]: yeah, but for this view, the isAssignedSK permission isn't assigned

File: server/audit/views.py
Comment (NicholasNagy) [reply server/audit/views.py:152]: yeah, but that's why we overrode the default implementation.

**Review by NicholasNagy — CHANGES_REQUESTED** [review]
Alright, so in terms of implementation/code on the front end it looks fine.

As for how the features look and feel, I do have a few comments.

1. Upon login, it should just show the audits. There should be no drawer on the side. On the top app bar, is should have the Alta logo instead of the hamburger menu and it should look like the following: 

image

When clicking on the Alta logo, it should bring you back to the audits page.

On the left of each audit, you shouldn't have the circles like you do in the app. They should either be like in the mockup above, or nonexistant to be implemented later.

Please also fix the colours in the background and list items to match what we have in the mockups.

2. The Bins should have a tab at the top to switch between the incompleted bins and the completed bins:

image

Please also fix the colours in the background and list items to match what we have in the mockups.

3. I think adding the camera and input buttons at the bottom were a great idea, but when you click on completed, it should disappear and only have a button for completing the bin. The list items should also look like in the following pic:

image

You should be able to confirm the item and flag it, but let's put that on hold for now, because looking at it from my screen, I feel like that would make it a little crammed.

For the completed items, as mentioned above, there should be the completed items button that when you click on, it completes the bin and puts the bin into the completed section. See the pic:
image

Also, we should have those circle indicators here to show which items were validated and missing. Missing items are not confirmed, so initially they are all missing and the user has to go and validate them all. New is for items not in the list of items to audit, which I think you've already caught onto.

For the settings button, I actually like that it gives a pop up instead of putting the 2 buttons for editing and removing. What you have here though is a delete button, it should be instead called remove, and instead of deleting the item, it should unscan it or invalidate it. If it's a new item though, the record can be deleted, but that is the only case when it can be deleted.

4. When editing an item, I can't exit the popup on my iphone. I realize through the browser it's possible by clicking outside the bounds of the popup, but it can't be done on my phone. Maybe an 'X' somewhere would fix it.

The extra information button at the top right (with the eye icon) should only show the info of the item and be ineditable. And the other page without the extra information button, should just display the following:

image

and when you click on the comments, it should open and close the text box. If the item is SERIALIZABLE, constrain the text box to quantities of only 1 or 0.


I think that covers everything I can think of right now for the UI. Once that's completed, I can look it over again and give feedback

File: server/user_account/permissions.py
Comment (Averynder) [reply server/user_account/permissions.py]: check the recent commits, I believe it addresses the security issue while still extending the functionality of the permission factory so it accommodates both sides

**Comment by Amanda-Kolopa** [comment]
Amazing work guys! :smile: 

I agree with the majority of Nick's review but I have some additional comments to add:

> On the left of each audit, you shouldn't have the circles like you do in the app. They should either be like in the mockup above, or nonexistant to be implemented later.

> image

As mentioned in the team meeting, Avery, Makram, and I had a discussion on the anticipated effort of some of the designs. They explained that the circular progress wheel may be difficult to incorporate and I had suggested switching to the color representations. If you don't approve of this, we can remove it for now and perhaps design something that's a bit more feasible. (same idea for the Bins page)

> I think adding the camera and input buttons at the bottom were a great idea, but when you click on completed, it should disappear and only have a button for completing the bin. The list items should also look like in the following pic:
> 
> image

What do you think about switching "Audit X" to "Bin X" since the back button would go back to the list of bins? 

> You should be able to confirm the item and flag it, but let's put that on hold for now, because looking at it from my screen, I feel like that would make it a little crammed.

In our discussion, it was mentioned that 

- when a stock-keeper scans a barcode (or the data is submitted through the camera or manually done), the item's status is set to _Provided_ and will automatically move to the _Completed Items_ list. 

- In the event where the stock-keeper has submitted a new item (they scanned an item which was on their list, some other item was added to the _Complete Items_ list, the original item is still in the _Items Left_ list), they can select the _Completed Items_ tab and then select the _Flag_ button for that item (as well comment on it if need be). This would use the popup window as expected in the mockups. 

- In the event where the stock-keeper cannot find the item (which is still found on the _Items Left_ page), the status is automatically set to _Missing_ and the corresponding popup window appears to add a comment and confirm (or just have an empty comment).

Currently, the popup window appears and allows the user to select the status of the item. We can have a quick group discussion to go over these designs and details! 

> For the settings button, I actually like that it gives a pop up instead of putting the 2 buttons for editing and removing. What you have here though is a delete button, it should be instead called remove, and instead of deleting the item, it should unscan it or invalidate it. If it's a new item though, the record can be deleted, but that is the only case when it can be deleted.

I don't think we should have the option to delete newly found items as it could misrepresent the inventory accuracy level at the end of the audit. Also, we want to keep track of discovered discrepancies to inform the inventory manager of overlooked items in their original inventory data.

> The extra information button at the top right (with the eye icon) should only show the info of the item and be ineditable. And the other page without the extra information button, should just display the following:
> 
> image

Maybe we can remove the item details in the popup as it provides too much information that may not be valuable to the stock-keeper. We can revisit this design with Ajay though.


Side note: please remove any lingering console.logs (ex. when selecting the _Input_ button and then selecting the _Cancel_ button)

**Comment by NicholasNagy** [comment]
> > On the left of each audit, you shouldn't have the circles like you do in the app. They should either be like in the mockup above, or nonexistant to be implemented later.
> 
> > image
> 
> As mentioned in the team meeting, Avery, Makram, and I had a discussion on the anticipated effort of some of the designs. They explained that the circular progress wheel may be difficult to incorporate and I had suggested switching to the color representations. If you don't approve of this, we can remove it for now and perhaps design something that's a bit more feasible. (same idea for the Bins page)

To be honest, I've done it before and it's not that difficult. I think the color representation of the completion % would be really vague and hard to understand, which defeats the purpose. If they find it too difficult, then we should just remove it, we can get much of the same info that the donut chart provides by just clicking on the info button.

> > I think adding the camera and input buttons at the bottom were a great idea, but when you click on completed, it should disappear and only have a button for completing the bin. The list items should also look like in the following pic:
> > image
> 
> What do you think about switching "Audit X" to "Bin X" since the back button would go back to the list of bins?

Sounds good!

> 
> > You should be able to confirm the item and flag it, but let's put that on hold for now, because looking at it from my screen, I feel like that would make it a little crammed.
> 
> In our discussion, it was mentioned that
> 
> * when a stock-keeper scans a barcode (or the data is submitted through the camera or manually done), the item's status is set to _Provided_ and will automatically move to the _Completed Items_ list.
> * In the event where the stock-keeper has submitted a new item (they scanned an item which was on their list, some other item was added to the _Complete Items_ list, the original item is still in the _Items Left_ list), they can select the _Completed Items_ tab and then select the _Flag_ button for that item (as well comment on it if need be). This would use the popup window as expected in the mockups.
> * In the event where the stock-keeper cannot find the item (which is still found on the _Items Left_ page), the status is automatically set to _Missing_ and the corresponding popup window appears to add a comment and confirm (or just have an empty comment).
> 
> Currently, the popup window appears and allows the user to select the status of the item. We can have a quick group discussion to go over these designs and details!

I agree with this analysis, however my comment was more related to when the barcode scanner is unable to function properly. They should be given the ability to manually validate each item through the interface.

> 
> > The extra information button at the top right (with the eye icon) should only show the info of the item and be ineditable. And the other page without the extra information button, should just display the following:
> > image
> 
> Maybe we can remove the item details in the popup as it provides too much information that may not be valuable to the stock-keeper. We can revisit this design with Ajay though.
> 

Yeah, I agree it gives too much info, but it gives them extra insight if they need it. We should discuss the details with Ajay however, on what they should and shouldn't be seeing (i.e. they shouldn't be seeing location, because that's useless for them, since they should know where they work)

**Comment by Amanda-Kolopa** [comment]
> To be honest, I've done it before and it's not that difficult. I think the color representation of the completion % would be really vague and hard to understand, which defeats the purpose. If they find it too difficult, then we should just remove it, we can get much of the same info that the donut chart provides by just clicking on the info button.

Yeah I agree, and sounds good! We can remove it and come back to it as a UI improvement.

> I agree with this analysis, however my comment was more related to when the barcode scanner is unable to function properly. They should be given the ability to manually validate each item through the interface.

Oh ok. Then yes I agree, in that sense, we should touch up the mockups to make it feel less crammed. 

> Yeah, I agree it gives too much info, but it gives them extra insight if they need it. We should discuss the details with Ajay however, on what they should and shouldn't be seeing (i.e. they shouldn't be seeing location, because that's useless for them, since they should know where they work)

Sounds good! I'll write it in our list of questions for our next stakeholder meeting.

**Comment by NicholasNagy** [comment]
> > To be honest, I've done it before and it's not that difficult. I think the color representation of the completion % would be really vague and hard to understand, which defeats the purpose. If they find it too difficult, then we should just remove it, we can get much of the same info that the donut chart provides by just clicking on the info button.
> 
> Yeah I agree, and sounds good! We can remove it and come back to it as a UI improvement.

Exactly!

> 
> > I agree with this analysis, however my comment was more related to when the barcode scanner is unable to function properly. They should be given the ability to manually validate each item through the interface.
> 
> Oh ok. Then yes I agree, in that sense, we should touch up the mockups to make it feel less crammed.
> 

Okay cool, I will try to look into it as soon as I get the chance. I think just removing the buttons would be the simplest way. I'll admit that having the confirm button is a lot simpler, but the more I think about it, the more I think it would be best for us to discourage making it too easy to confirm an item with a click of a button. This is simply because if I were in their place, I would just look through them and keep clicking, and that would be prone to making more errors than necessary, since the rate at which you would be able to confirm would be way too high. The scanner forces technology to validate for them, which isn't very error prone.

We can take this offline and discuss if you prefer?

> > Yeah, I agree it gives too much info, but it gives them extra insight if they need it. We should discuss the details with Ajay however, on what they should and shouldn't be seeing (i.e. they shouldn't be seeing location, because that's useless for them, since they should know where they work)
> 
> Sounds good! I'll write it in our list of questions for our next stakeholder meeting.

Thank you!

**Comment by Amanda-Kolopa** [comment]
> Okay cool, I will try to look into it as soon as I get the chance. I think just removing the buttons would be the simplest way. I'll admit that having the confirm button is a lot simpler, but the more I think about it, the more I think it would be best for us to discourage making it too easy to confirm an item with a click of a button. This is simply because if I were in their place, I would just look through them and keep clicking, and that would be prone to making more errors than necessary, since the rate at which you would be able to confirm would be way too high. The scanner forces technology to validate for them, which isn't very error prone.
> 
> We can take this offline and discuss if you prefer?

You have a valid point. I think we should discuss the details offline to not lengthen the PR comments. In terms of UI, I think we can leave it for now and make the improvements once the PR is merged. The mockups should be adjusted by then.

**Review by NicholasNagy — APPROVED** [review]
Thank you for all the changes, it now looks good to merge.

**Pull request merged**
